sled-agent/src/rack_setup/plan/service.rs

```diff
@@ -230,224 +233,322 @@ impl Plan {
         storage: &StorageResources,
         sleds: &HashMap<SocketAddrV6, StartSledAgentRequest>,
     ) -> Result<Self, PlanError> {
-        let reserved_rack_subnet = ReservedRackSubnet::new(config.az_subnet());
-        let dns_subnets = reserved_rack_subnet.get_dns_subnets();
-
-        let mut allocations = vec![];
         let mut dns_builder = internal_dns::DnsConfigBuilder::new();
-
-        let rack_dns_servers = dns_subnets
-            .clone()
-            .into_iter()
-            .map(|dns_subnet| dns_subnet.dns_address().ip().to_string())
-            .collect::<Vec<String>>();
-
         let mut services_ip_pool = config
             .internal_services_ip_pool_ranges
             .iter()
             .flat_map(|range| range.iter());
+        let mut svc_port_builder = ServicePortBuilder::new();
 
-        let mut boundary_ntp_servers = vec![];
-        let mut seen_any_scrimlet = false;
+        // Load the information we need about each Sled to be able to allocate
+        // components on it.
+        let mut sled_info = {
+            let result: Result<Vec<SledInfo>, PlanError> =
+                futures::future::try_join_all(sleds.values().map(
+                    |sled_request| async {
+                        let subnet = sled_request.subnet;
+                        let sled_address = get_sled_address(subnet);
+                        let u2_zpools =
+                            Self::get_u2_zpools_from_sled(log, sled_address)
+                                .await?;
+                        let is_scrimlet =
+                            Self::is_sled_scrimlet(log, sled_address).await?;
+                        Ok(SledInfo::new(
+                            sled_request.id,
+                            subnet,
+                            sled_address,
+                            u2_zpools,
+                            is_scrimlet,
+                        ))
+                    },
+                ))
+                .await;
+            result?
+        };
```

Comment:
Should we just make this `Vec<SledInfo>` a required input to this function?

If we do so, I don't *think* this function tries to contact sled agents directly, and it might be easier to write tests using a "TransientDnsServer"

Reply:
Yes, I think that's a great idea.